src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceGenerator.cs

```diff
@@ -58,283 +50,149 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
+
+            var interfacesToGenerate = interfaceSymbolsWithoutDiagnostics
+                .Select((data, ct) => data.InterfaceInfo!);
+
+            var interfaceContexts = interfacesToGenerate.Collect().SelectMany(ComInterfaceContext.GetContexts);
+
+            // Get the information we need about methods themselves
+            var interfaceMethodsAndDiagnostics = interfaceSymbolsWithoutDiagnostics.Select(ComMethodInfo.GetMethodsFromInterface);
+            context.RegisterDiagnostics(interfaceMethodsAndDiagnostics.SelectMany(static (methodList, ct) => methodList.Select(m => m.Diagnostic)));
+            var interfaceMethods = interfaceMethodsAndDiagnostics
+                .Select(static (methods, ct) =>
+                    methods
+                        .Where(pair => pair.Diagnostic is null)
+                        .Select(pair => pair.ComMethod)
+                        .ToSequenceEqualImmutableArray());
+
+            // Generate a map from Com interface to the methods it declares
+            var interfaceToDeclaredMethodsMap = interfaceContexts
+                .Zip(interfaceMethods)
+                .Collect()
                 .Select(static (data, ct) =>
-            {
-                var (interfaceData, interfaceContext) = data;
-                Location interfaceLocation = interfaceData.Syntax.GetLocation();
-                var methods = ImmutableArray.CreateBuilder<(MethodDeclarationSyntax Syntax, IMethodSymbol Symbol, int Index, Diagnostic? Diagnostic)>();
-                int methodVtableOffset = interfaceContext.MethodStartIndex;
-                foreach (var member in interfaceData.Symbol.GetMembers())
                 {
-                    if (member.Kind == SymbolKind.Method && !member.IsStatic)
-                    {
-                        // We only support methods that are defined in the same partial interface definition as the
-                        // [GeneratedComInterface] attribute.
-                        // This restriction not only makes finding the syntax for a given method cheaper,
-                        // but it also enables us to ensure that we can determine vtable method order easily.
-                        Location? locationInAttributeSyntax = null;
-                        foreach (var location in member.Locations)
-                        {
-                            if (location.SourceTree == interfaceLocation.SourceTree
-                                && interfaceLocation.SourceSpan.Contains(location.SourceSpan))
-                            {
-                                locationInAttributeSyntax = location;
-                            }
-                        }
+                    return data.ToValueEqualityImmutableDictionary<(ComInterfaceContext, SequenceEqualImmutableArray<ComMethodInfo>), ComInterfaceContext, SequenceEqualImmutableArray<ComMethodInfo>>(
+                        static pair => pair.Item1,
+                        static pair => pair.Item2);
```

Comment:
This uses a wrapper type around `ImmutableDictionary` (`ValueEqualImmutableDictionary`) which uses set equality on the inner dictionary, so I think it should still be incremental, but it does allocate for the comparison. I can change it to something like that, though. It'll still need to collect everything to get all the inherited methods for each type, but it might be better than the dictionary as the incremental value.

Reply:
After a little refactoring, we don't need this dictionary anymore.